File: io/json/src/test/resources/compat/issue-756/src/java/LegacyJsonStreamCaller.java
package io.bluetape4k.json.compat.issue756.java;

import io.bluetape4k.json.JsonSerializer;
import java.io.ByteArrayOutputStream;
import java.io.OutputStream;
import java.lang.reflect.Method;
import java.nio.charset.StandardCharsets;

public final class LegacyJsonStreamCaller {
    private LegacyJsonStreamCaller() {
    }

    public static void main(String[] args) throws Exception {
        JsonSerializer serializer = new LegacyJsonImplementation();
        require("json".equals(new String(serializer.serialize("json"), StandardCharsets.UTF_8)),
            "legacy JSON call changed");

        Method method = JsonSerializer.class.getMethod(
            "serializeJsonToStream",
            Object.class,
            OutputStream.class
        );
        require(method.isDefault(), "JSON stream method is not a default");

        ByteArrayOutputStream target = new ByteArrayOutputStream();
        Object written = method.invoke(serializer, "json", target);
        require(Integer.valueOf(4).equals(written), "unexpected JSON stream write count");
        require("json".equals(target.toString(StandardCharsets.UTF_8)), "unexpected JSON stream payload");
        System.out.println("legacy-json-java-stream-caller=PASS");
    }

    private static void require(boolean condition, String message) {
        if (!condition) {
            throw new AssertionError(message);
        }
    }
}

File: io/json/src/test/resources/compat/issue-756/src/kotlin/LegacyJsonStreamCaller.kt
package io.bluetape4k.json.compat.issue756.kotlin

import io.bluetape4k.json.JsonSerializer
import java.io.ByteArrayOutputStream
import java.io.OutputStream

private class LegacyJsonKotlinImplementation: JsonSerializer {
    override fun serialize(graph: Any?): ByteArray = graph?.toString()?.encodeToByteArray() ?: byteArrayOf()

    override fun <T: Any> deserialize(bytes: ByteArray?, clazz: Class<T>): T? =
        bytes?.takeIf { it.isNotEmpty() }?.decodeToString()?.let(clazz::cast)
}

fun main() {
    val serializer: JsonSerializer = LegacyJsonKotlinImplementation()
    check(serializer.serialize("json").decodeToString() == "json") { "legacy JSON call changed" }

    val method = JsonSerializer::class.java.getMethod(
        "serializeJsonToStream",
        Any::class.java,
        OutputStream::class.java,
    )
    check(method.isDefault) { "JSON stream method is not a default" }

    val target = ByteArrayOutputStream()
    val written = method.invoke(serializer, "json", target)
    check(written == 4) { "unexpected JSON stream write count" }
    check(target.toString(Charsets.UTF_8) == "json") { "unexpected JSON stream payload" }
    println("legacy-json-kotlin-stream-caller=PASS")
}
